File: 二叉树/144.py
'''
    二叉树的前序遍历
'''


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# 递归
class Recursion:
    def preorderTraversal(self, root: TreeNode) -> list:
        res = []
        self.preorder(root, res)
        return res

    def preorder(self, root, res):
        if(root):
            res.append(root.val)
            self.preorder(root.left, res)
            self.preorder(root.right, res)


# 递归
class Iteration:
    def preorderTraversal(self, root: TreeNode) -> list:
        res = []
        stack = []
        if(root):
            stack.append(root)
        while(stack):
            node = stack.pop()
            res.append(node.val)
            if(node.right):
                stack.append(node.right)
            if(node.left):
                stack.append(node.left)
        return res

root = TreeNode(0)
root.left = TreeNode(1)
root.right = TreeNode(2)
root.left.left = TreeNode(3)
root.left.right = TreeNode(4)
root.right.left = TreeNode(5)
root.right.right = TreeNode(6)
obj = Iteration()
print(obj.preorderTraversal(root))

File: 二叉树/144.cpp
#include <vector>
#include <iostream>
using namespace std;

//Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// 递归
class Recursion
{
public:
    vector<int> res;
    vector<int> preorderTraversal(TreeNode *root)
    {
        if (root)
        {
            res.push_back(root->val);
            preorderTraversal(root->left);
            preorderTraversal(root->right);
        }
        return res;
    }
};

// 迭代
class Iteration
{
public:
    vector<int> preorderTraversal(TreeNode *root)
    {
        if(!root)
            return {};
        vector<int> res;
        vector<TreeNode *> NodeStack;
        NodeStack.push_back(root);
        while (!NodeStack.empty())
        {
            TreeNode *node = NodeStack.back();
            res.push_back(node->val);
            NodeStack.pop_back();
            if (node->right)
                NodeStack.push_back(node->right);
            if (node->left)
                NodeStack.push_back(node->left);
        }
        return res;
    }
};

int main()
{
    TreeNode *root = new TreeNode(0);
    root->left = new TreeNode(1);
    root->left->left = new TreeNode(3);
    root->left->right = new TreeNode(4);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(5);
    root->right->right = new TreeNode(6);
    Iteration obj;
    vector<int> res = obj.preorderTraversal(root);
    for (auto num : res)
        cout << num << endl;
}